executor/src/witgen/mod.rs

```diff
@@ -49,14 +49,37 @@ static OUTER_CODE_NAME: &str = "witgen (outer code)";
 pub trait QueryCallback<T>: Fn(&str) -> Result<Option<T>, String> + Send + Sync {}
 impl<T, F> QueryCallback<T> for F where F: Fn(&str) -> Result<Option<T>, String> + Send + Sync {}
 
+type WitgenCallbackFn<T> =
+    Arc<dyn Fn(&[(String, Vec<T>)], BTreeMap<u64, T>, u8) -> Vec<(String, Vec<T>)> + Send + Sync>;
+
 #[derive(Clone)]
-pub struct WitgenCallback<T> {
+pub struct WitgenCallback<T>(WitgenCallbackFn<T>);
+
+impl<T: FieldElement> WitgenCallback<T> {
+    pub fn new(f: WitgenCallbackFn<T>) -> Self {
+        WitgenCallback(f)
+    }
+
+    /// Computes the next-stage witness, given the current witness and challenges.
+    pub fn next_stage_witness(
+        &self,
+        current_witness: &[(String, Vec<T>)],
+        challenges: BTreeMap<u64, T>,
+        stage: u8,
+    ) -> Vec<(String, Vec<T>)> {
+        (self.0)(current_witness, challenges, stage)
+    }
+}
+
+pub struct WitgenCallbackContext<T> {
+    /// TODO: all these fields probably don't need to be Arc anymore, since the
+    /// Arc was moved one level up... but I have to investigate this further.
```

Comment:
Lets leave this TODO to a next PR (that I am already working on).

Reply:
I give up. It goes too deep. I still don't understand the need for backend to take `Arc` instead of plain references, though.

Reply:
That was me :P #1480

> Receiving references, the `CompositeBackend` has no chance of changing the PIL or fixed columns before passing it to the wrapped backend. Using `Arc` fixed it, and is quite compatible with what is kept by `Pipeline` already.

I think a `Cow` would have worked as well?

Reply:
But Arc is also immutable. How does it make easier than using reference?

Reply:
But you can cheaply clone. So, even though the prove function "owns" the witness, in most cases, it is really just a wrapped reference. But the `CompositeBackend` can actually deep clone the witness and modify it (split by machine) and create a new `Arc`.

I don't think it would be possible with a reference. See `BackendFactory::create`:
https://github.com/powdr-labs/powdr/blob/e58ddac313963f7c4a57f2ad4760b1f54eba6791/backend/src/lib.rs#L131-L143

The returned backend needs to either own its data or hold a reference who's lifetime outlives `BackendFactory::create`. So `CompositeBackendFactory::create` couldn't create a new `Analyzed<T>` and pass a reference to the backend.

Reply:
Well, the first thing you do is:
```
        let pils = split::split_pil((*pil).clone());
```
and then you don't touch the original `pil` again. That would work just fine if it was a reference, no?

You create new `Arc<Analyzed<F>>` for the `CompositeBackend` object (that could probably even be `Box<Analyzed<F>>`), but that is irrelevant for the `Backend` API.

All that said, I understand that it is more comfortable and flexible to use `Arc` and the cost is negligible, and if I would try to revert this change, it is possible that soon in the future someone would want to change it back to `Arc` again, so I am not saying we should revert it. But for this case, I still don't see why it was needed in the first place...

PS: now that we use `Arc`, the lifetime `'a` in the `Backend` trait is not needed anymore.
PS2: `split::split_pil` should probably take a reference